### DISCUSSION TITLE
Creating a 2D polycrystal with a grain size gradient

### DISCUSSION THREAD

**Discussion by ehsanmoshkelgosha22** [neperfepx/neper#228] [Q&A] [answered]
Hello Romain,

I wonder how it is possible to generate a 2D polycrystal with different grain size distributions similar to the below picture.

![4](https://user-images.githubusercontent.com/49301391/119428666-912d1a80-bcca-11eb-92c6-fdf5e18013fa.jpg)

Thanks,
Ehsan

**Comment by rquey** [comment, marked as answer]
There is no metric for this in `-morpho` at the moment. The only option is to generate and load a custom seed distribution, with a gradient of seed density.

Related issue: #101.

**Reply by ehsanmoshkelgosha22** [reply]
Thanks for your reply.

**Reply by ehsanmoshkelgosha22** [reply]
I did not get your point clearly. Could you please explain how it is possible to load a custom seed distribution with a gradient of seed density?

**Reply by rquey** [reply]
In simple terms, if you put more seeds on the right than on the left, you will get the size gradient. You can load a custom seed distribution using `-morphooptiini "coo:file(...)"`.

**Reply by smallrain1996** [reply]
![image](https://github.com/neperfepx/neper/assets/52028117/916806c3-a9fc-4557-9486-acd0fe7cc622)
I tried this method successfully.

**Reply by zengyan0112** [reply]
Can you give specific orders？ Thank you very much.

**Comment by changhaoliu-UOM** [comment]
Hello everyone in this discussion,

@zengyan0112 @smallrain1996  I currently need the gradient structure as you mentioned. 2D and 3D structures are both necessary, so could you please give some recommendations and if you can share some codes to build, that would be fantastic.

Thank you very much.

best
Changhao

**Reply by rquey** [reply]
See https://github.com/neperfepx/neper/discussions/228#discussioncomment-14245494.

**Comment by Parthasarathi200127** [comment]
Hello everyone

@rquey i understood what you are saying but how can i put more seeds where it is required
i also want to generate this type of microstructure
<img width="155" height="532" alt="Screenshot (17)" src="https://github.com/user-attachments/assets/d449adb1-d02a-4327-b5ed-d2f65224b0c3" />

**Reply by rquey** [reply]
Here is an example:
```
gsl-randist 1 1000 flat 0 1 > x
gsl-randist 2 1000 flat 0 1 | awk '{print  4*$1*$1*$1}' > y
paste x y > coo
neper -T -dim 2 -n 1000 -morphooptiini "coo:file(coo)" -domain "square(1,4)" -o gradient
neper -V gradient.tess -print d
```
<img width="1200" height="900" alt="d" src="https://github.com/user-attachments/assets/b3dc36eb-005b-46be-91fa-e6d0a900d5f2" />

Adjust to your case.
